### PR TITLE
D42 16383 - Ansible inventory plugin timeout

### PR DESCRIPTION
**Pull request by osanchez42**

### CODE REVIEW

File: plugins/inventory/d42.py [hunk 1]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
                 self._add_host_to_keyed_groups(self.get_option('keyed_groups'), dict(), host_name, strict)
         except Exception as e:
             print(e)
+
+    def get_doql_csv(self, query):
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: Should probably rename this get_doql_json.

File: plugins/inventory/d42.py [hunk 2]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+        base_url = self.get_option('url')
+        username = self.get_option('username')
+        password = self.get_option('password')
+        ssl_check = self.get_option('ssl_check')
+        debug = self.get_option('debug')
+
+        data = {'output_type': 'json', 'header': 'yes', 'query': query}
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: I don't think header param is needed for json output.

File: plugins/inventory/d42.py [hunk 3]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+        try:
+            # while there should be no timeout, ansible seems to get stuck sending requests without timeouts
+            response = requests.post(base_url + "/services/data/v1.0/query/", data=data,
+                                    auth=(username, password), verify=ssl_check, timeout=30)
+
+            if debug:
+                status_code = response.status_code
+                print('Response Status: ' + str(status_code))
+
+            # csv response to json object
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: Update comment.  You now have a json response.

File: plugins/inventory/d42.py [hunk 4]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+        for device in d42_devices:
+
+            device_record = {}
+
+            device_record['datastores'] = device.get('datastores')
+            device_record['last_updated'] = device.get('last_edited')
+            device_record['ip_addresses'] = []
+            device_record['serial_no'] = device.get('serial_no')
+            device_record['hw_depth'] = device.get('hw_depth')
+            device_record['nonauthoritativealiases'] = str(device.get('nonauthoritativealiases')).split(',')
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: What happens if the device has no nonauthoritativealiases?  It looks like if None came back, then we would end up with the array ['None'].

File: plugins/inventory/d42.py [hunk 5]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+            device_record['is_it_blade_host'] = 'no' if device.get('blade_chassis') == 'false' else 'yes'
+            device_record['is_it_switch'] = 'no' if device.get('is_it_switch') == 'false' else 'yes'
+            device_record['virtual_subtype_id'] = device.get('virtual_subtype_id')
+            device_record['hw_size'] = device.get('hw_size')
+            device_record['id'] = device.get('device_pk')
+
+            # custom field
+            device_record['custom_fields'] = []
+
+            device_record['aliases'] = str(device.get('aliases')).split(',')
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: What is the result if the device has no aliases?

File: plugins/inventory/d42.py [hunk 6]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+
+            device_record['datastores'] = device.get('datastores')
+            device_record['last_updated'] = device.get('last_edited')
+            device_record['ip_addresses'] = []
+            device_record['serial_no'] = device.get('serial_no')
+            device_record['hw_depth'] = device.get('hw_depth')
+            device_record['nonauthoritativealiases'] = str(device.get('nonauthoritativealiases')).split(',')
+
+            device_record['service_level'] = device.get('service_level')
+            device_record['is_it_blade_host'] = 'no' if device.get('blade_chassis') == 'false' else 'yes'
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:195]: To confirm, we get back a string representation of a Boolean and not an actual Boolean?

File: plugins/inventory/d42.py [hunk 7]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+            if device_id in d42_inventory:
+                d42_inventory[device_id]['hdd_details'].append({
+                    "raid_group": hdd_detail_record.get('raid_group'),
+                    "serial_no": hdd_detail_record.get('hdd_serial_no'),
+                    "raid_type": hdd_detail_record.get('hdd_raid_type'),
+                    "hdd": {
+                        "description": hdd_detail_record.get('hdd_description'),
+                        "partno": hdd_detail_record.get('partno'),
+                        "rpm": {
+                            "id": hdd_detail_record.get('hdd_rom_id'),
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: Did you mean for this to be spelled hdd_rpm_id?

File: plugins/inventory/d42.py [hunk 8]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+                    {
+                        "subnet_id": device_ip_record.get('subnet_id'),
+                        "ip": device_ip_record.get('ip'),
+                        "label": device_ip_record.get('label'),
+                        "type": device_ip_record.get('type'),
+                        "subnet": device_ip_record.get('subnet')
+                    }
+                )
+            else:
+                print(device_id + ' Not found in d42_inventory')
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py]: Was this print for debugging purposes?

File: plugins/inventory/d42.py [hunk 9]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+
+    def get_external_links(self):
+        external_links_query = """
+                select
+                view_device_v1.device_pk,
+                view_device_v1.name as device_name,
+                external_links.device_url,
+                external_links.device_url_notes
+                from view_device_v1
+                inner join (select device_fk, device_url, notes as device_url_notes from view_deviceurl_v1) external_links
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:518]: Do you even need the join on view_device_v1 since view_deivceurl_v1 has the device_fk?

File: plugins/inventory/d42.py [hunk 10]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+            select
+            view_device_v1.device_pk,
+            view_device_v1.name,
+            ip_address.subnet_fk as subnet_id,
+            ip_address.ip_address as ip,
+            ip_address.label,
+            ip_address.type,
+            subnet.name as subnet
+            from view_device_v1
+            left join (select device_fk, subnet_fk, ip_address, label, type from view_ipaddress_v1) ip_address
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:568]: Do you need view_device_v1 since view_ipaddress_v1 already has device_fk?

File: plugins/inventory/d42.py [hunk 11]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+                CASE
+                    WHEN length(TRIM(hwaddress)) = 12
+                        THEN CONCAT_WS(':',left(TRIM(view_netport_v1.hwaddress),2), substring(TRIM(view_netport_v1.hwaddress),3,2), substring(TRIM(view_netport_v1.hwaddress),5,2), substring(TRIM(view_netport_v1.hwaddress),7,2), substring(TRIM(view_netport_v1.hwaddress),9,2), right(TRIM(view_netport_v1.hwaddress),2))
+                        ELSE
+                            hwaddress
+                        END
+                )
+                as hw_address, name as port_name, port, primary_vlan_fk from view_netport_v1) mac_address
+            ON view_device_v1.device_pk = mac_address.device_fk
+            full join (select vlan_pk, name from view_vlan_v1) vlan
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:595]: Couldn't you do a left outer join to view_vlan_v1, why did you need a full join?

File: plugins/inventory/d42.py [hunk 12]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+    def get_mac_addresses(self):
+        mac_addresses_query = """
+            select
+            view_device_v1.device_pk,
+            view_device_v1.name as device_name,
+            mac_address.hw_address as mac,
+            mac_address.port_name,
+            mac_address.port,
+            vlan.name as vlan
+            from view_device_v1
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:584]: Do you need view_device_v1 since view_netport_v1 has device_fk?

File: plugins/inventory/d42.py [hunk 13]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+            line_items.frequency as line_frequency,
+            line_items.cost as line_cost,
+            line_items.cancel_policy as line_cancel_policy,
+            line_items.item_type as line_item_type,
+            purchase.order_no as purchase_order_no,
+            line_items.line_type as line_type,
+            line_items.service_type_name as line_service_type,
+            line_items.end_date as line_end_date,
+            line_items.contract_type_name as line_contract_type
+            from view_device_v1
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:623]: Don't think we need view_device_v1 since the other view already has device_fk.

File: plugins/inventory/d42.py [hunk 14]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+    def get_custom_fields(self):
+
+        custom_field_query = """
+                select
+                view_device_v1.device_pk,
+                view_device_v1.name,
+                custom_field.key as custom_field_key,
+                custom_field.value as custom_field_value,
+                custom_field.notes as custom_field_notes
+                from view_device_v1
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:492]: I don't think the join to view_device_v1 is needed since view_device_custom_fields has device_fk.

File: plugins/inventory/d42.py [hunk 15]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+        hdd.location as hdd_location,
+        hdd.hdd_id,
+        hdd.manufactuerer_id,
+        hdd.hdd_type_id,
+        hdd.hdd_type_name,
+        hdd.size,
+        hdd_details.part_id,
+        hdd_details.hddcount,
+        hdd_details.description
+        from view_device_v1
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:547]: Do you even need the join on view_device_v1 since view_part_v1 has the device_fk?

File: plugins/inventory/d42.py [hunk 16]
```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
...
+                                                THEN view_device_v1.hard_disk_size * 1024 * 1024
+                                            ELSE
+                                                NULL
+                                        END
+                                ELSE NULL
+                            END)) AS disk_size
+            from view_device_v1
+            left join view_hardware_v1
+                on view_device_v1.hardware_fk = view_hardware_v1.hardware_pk
+            left join  (select view_ipaddress_v1.device_fk,  string_agg(host(view_ipaddress_v1.ip_address)::character varying, ', ' order by view_ipaddress_v1.ip_address) as ip_address from view_ipaddress_v1 group by view_ipaddress_v1.device_fk) network_ip
```
Comment (cscaglioned42) [inline plugins/inventory/d42.py:434]: Are the joins to view_ipaddress_v1, view_netport_v1, view_purchaselineitem_v1 still needed since separate queries are being used to get them?

File: plugins/inventory/d42.py (hunk 4 above)
Comment (osanchez42) [reply plugins/inventory/d42.py]: correct, returns ['None'], result should just be []. fixing this now

File: plugins/inventory/d42.py (hunk 5 above)
Comment (osanchez42) [reply plugins/inventory/d42.py]: returns ['None'], result should just be []. fixing this now

File: plugins/inventory/d42.py (hunk 6 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:195]: yes, that is how /api/1.0/devices/all returns these values

File: plugins/inventory/d42.py (hunk 8 above)
Comment (osanchez42) [reply plugins/inventory/d42.py]: yes

File: plugins/inventory/d42.py (hunk 7 above)
Comment (osanchez42) [reply plugins/inventory/d42.py]: mispelled in both query in query too, should not affect results, but changing now to say rpm

File: plugins/inventory/d42.py (hunk 9 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:518]: I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

Comment (osanchez42) [reply plugins/inventory/d42.py:520]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 10 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:568]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 11 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:595]: if vlan field is empty I do not get any records with left outer join. this makes up for empty vlan field so that if a mac address does not have a vlan it is still included

File: plugins/inventory/d42.py (hunk 12 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:584]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 13 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:623]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 14 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:492]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 15 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:547]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 16 above)
Comment (osanchez42) [reply plugins/inventory/d42.py:434]: > I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device

File: plugins/inventory/d42.py (hunk 11 above)
Comment (cscaglioned42) [reply plugins/inventory/d42.py:595]: Left outer join should work.  I think a full join would return rows where vlan exists, but the mac address does not.
